src/timesync/timesyncd-manager.c

```diff
@@ -1130,3 +1137,50 @@ int manager_new(Manager **ret) {
 
         return 0;
 }
+
+static int manager_save_time_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+        Manager *m = userdata;
+
+        assert(m);
+
+        return manager_save_time_and_rearm(m);
+}
+
+int manager_save_time_add(Manager *m) {
+        int r;
+
+        if (m->save_time_interval_usec > 0) {
+                r = sd_event_add_time_relative(
+                                m->event, &m->event_save_time,
+                                clock_boottime_or_monotonic(),
+                                m->save_time_interval_usec, 0, manager_save_time_handler, m);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to add save time event: %m");
+
+                r = sd_event_source_set_enabled(m->event_save_time, SD_EVENT_ONESHOT);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to enable save time event: %m");
+        }
+
+        return 0;
+}
+
+static int manager_save_time_and_rearm(Manager *m) {
+        int r;
+
+        r = touch(CLOCK_FILE);
+        if (r < 0)
+                log_debug_errno(r, "Failed to update %s, ignoring: %m", CLOCK_FILE);
+
+        if (m->save_time_interval_usec > 0) {
+                r = sd_event_source_set_time_relative(m->event_save_time, m->save_time_interval_usec);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to rearm save time event: %m");
+
+                r = sd_event_source_set_enabled(m->event_save_time, SD_EVENT_ONESHOT);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to enable save time event: %m");
+        }
```

Comment:
hmm, ther's no code to free the timer event again on shutdown

Reply:
:facepalm: Added.